# FUNNEL SHEET ANSWERS

## Our objective is to increase the LV5/LV1 ratio.

In [1]:
# import the basic required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 

In [3]:
# read our dataset 
df = pd.read_csv('fittlyf_data.csv')

### Now try to explore the dataset and see what our dataset tells us

### EDA and Statistical Analysis

In [4]:
df.shape

(144, 6)

We had **144** rows and **6** columns in our dataset, we can say it is small dataset.

In [5]:
# just see how our dataset looks like
df.head()

Months (Date)        Date   Variants DeviceType  Sum of Clicks  \
0           Mar  2023/03/29    Control    Desktop          31312   
1           Mar  2023/03/29    Control     Mobile          18399   
2           Mar  2023/03/29    Control     Others             34   
3           Mar  2023/03/29    Control     Tablet            660   
4           Mar  2023/03/29  Treatment    Desktop          10067   

   Sum of Visitors  
0           298032  
1           192144  
2              375  
3             7646  
4            32737

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144 entries, 0 to 143
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Months (Date)    144 non-null    object
 1   Date             144 non-null    object
 2   Variants         144 non-null    object
 3   DeviceType       144 non-null    object
 4   Sum of Clicks    144 non-null    int64 
 5   Sum of Visitors  144 non-null    int64 
dtypes: int64(2), object(4)
memory usage: 6.9+ KB


We can see there are **144** rows of every column is non-null.<br>
So,we can say there are **no missing values**, no need to handle the missing data. <br>
We can also observe here there are **4 categorical columns** and **2 numerical columns**<br>


In [7]:
# to check some statistics of the numerical data we have.
df.describe()

Sum of Clicks  Sum of Visitors
count     144.000000       144.000000
mean     6961.159722     60134.881944
std      9533.789003     93239.709621
min         1.000000        23.000000
25%        83.250000       574.250000
50%      1488.000000      9188.500000
75%     11253.500000     58703.750000
max     34476.000000    313539.000000

#### To check missing values.

In [8]:
# To check missing values
df.isnull().sum()

Months (Date)      0
Date               0
Variants           0
DeviceType         0
Sum of Clicks      0
Sum of Visitors    0
dtype: int64

No missing values.

In [9]:
# To check different unique values of variants with their frequency.
df.Variants.value_counts()

Control      72
Treatment    72
Name: Variants, dtype: int64

We can see that **control** and **treatment** have both same weightage.

In [10]:
# To check different unique values of DeviceType with their frequency.
df.DeviceType.value_counts()

Desktop    36
Mobile     36
Others     36
Tablet     36
Name: DeviceType, dtype: int64

One thing I can observe in dataset, mostly uniques values are divided **equally.**

Now we will see when does **sum of clicks** is high like when the variant is in **control state** or **treatment state**. <br>
We compare it by average.

In [14]:
# Calculate average sum of clicks for Control and Treatment
average_clicks_control = df[df['Variants'] == 'Control']['Sum of Clicks'].mean()
average_clicks_treatment = df[df['Variants'] == 'Treatment']['Sum of Clicks'].mean()

print(f'Average Clicks - Control: {average_clicks_control}')
print(f'Average Clicks - Treatment: {average_clicks_treatment}')


Average Clicks - Control: 11100.513888888889
Average Clicks - Treatment: 2821.8055555555557


We can observe that when it is **control state**, tendenct to click is **high**.

Similarly we will do for **sum of visitors**

In [15]:
# Calculate average sum of visitors for Control and Treatment
average_visitors_control = df[df['Variants'] == 'Control']['Sum of Visitors'].mean()
average_visitors_treatment = df[df['Variants'] == 'Treatment']['Sum of Visitors'].mean()

print(f'Average Visitors - Control: {average_visitors_control}')
print(f'Average Visitors - Treatment: {average_visitors_treatment}')

Average Visitors - Control: 108312.375
Average Visitors - Treatment: 11957.388888888889


We can come to conclusion that when it is in **control state**, high chances of visitors.

Once we will look at **time** related cols also. like **month and date**.

In [24]:
# Check unique values of 'Months (Date)' with their frequency
unique_months_frequency = df['Months (Date)'].value_counts()

# Print the result
print(unique_months_frequency)

Apr    120
Mar     24
Name: Months (Date), dtype: int64


Mostly this data is taken or under observation in **April** month.

**EDA Completed**

## Question (1): Look at the data and replace the blank values in the ‘value’ column with logical values. 

From above **EDA** part only, we can say we don't have any missing values, so don't need to handle them.

## Question (2): Create a function which takes region, Customer segment, Start Year & Month, End year & month as an input parameter and gives the following as an output.

#### 2(a): A line graph & a bar graph to show the trend of KPIs column for the given date range (i.e., between the Start Year & Month and End year & month). The graph should have appropriate labels, titles and other factors which would make it readable. 

In [30]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [33]:
def plot_kpis_trend(region, start_year, start_month, end_year, end_month):
    # Extract start and end dates
    start_date = f"{start_year}/{start_month}/01"
    end_date = f"{end_year}/{end_month}/31"

    # Filter the data based on the 'Months (Date)' column
    filtered_data = df[(df['Months (Date)'] >= start_date) & (df['Months (Date)'] <= end_date) & (df['DeviceType'] == region)]

    # Check if filtered_data is not empty
    if not filtered_data.empty:
        # Plotting
        plt.figure(figsize=(12, 6))

        # Line Graph
        plt.subplot(1, 2, 1)
        sns.lineplot(x='Months (Date)', y='Sum of Clicks', hue='Variants', data=filtered_data, marker='o')
        plt.title('Trend of Clicks Over Time')
        plt.xlabel('Months (Date)')
        plt.ylabel('Sum of Clicks')
        plt.xticks(rotation=45)

        # Bar Graph
        plt.subplot(1, 2, 2)
        sns.barplot(x='Months (Date)', y='Sum of Visitors', hue='Variants', data=filtered_data)
        plt.title('Bar Chart of Visitors Over Time')
        plt.xlabel('Months (Date)')
        plt.ylabel('Sum of Visitors')
        plt.xticks(rotation=45)

        plt.tight_layout()
        plt.show()
    else:
        print('No data available for the specified filters.')


In [35]:
# Example usage:
plot_kpis_trend(region='Desktop', start_year='2023', start_month='03', end_year='2023', end_month='04')

No data available for the specified filters.


In [36]:
# Example usage:
plot_kpis_trend(region='Mobile', start_year='2023', start_month='04', end_year='2023', end_month='04')


No data available for the specified filters.


In [37]:
# Example usage:
plot_kpis_trend(region='Desktop', start_year='2023', start_month='04', end_year='2023', end_month='04')


No data available for the specified filters.


#### 2(b) :Based on the input parameters, forecasts the Value column using the following algorithm and spit out the actual values and the forecasted values appended in a single csv file for each algorithm:<br>  i.	SARIMA <br>ii.	Long Short-Term Memory (LSTM) Networks<br>iii.	Moving average


In [39]:
!pip install pandas numpy matplotlib seaborn statsmodels tensorflow scikit-learn


In [40]:
# Required Libraries
import pandas as pd
import numpy as np
from statsmodels.tsa.statespace.sarimax import SARIMAX
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
import matplotlib.pyplot as plt

In [41]:
def forecast_and_save(region, start_year, start_month, end_year, end_month):
    # Extract start and end dates
    start_date = f"{start_year}/{start_month}/01"
    end_date = f"{end_year}/{end_month}/31"

    # Filter the data based on the 'Months (Date)' column
    filtered_data = df[(df['Months (Date)'] >= start_date) & (df['Months (Date)'] <= end_date) & (df['DeviceType'] == region)]

    # Check if filtered_data is not empty
    if not filtered_data.empty:
        # Extract relevant columns
        time_series_data = filtered_data[['Months (Date)', 'Sum of Visitors']].rename(columns={'Months (Date)': 'Date', 'Sum of Visitors': 'Value'})

        # Save the original time series data to a CSV file
        time_series_data.to_csv('original_data.csv', index=False)

        # Forecast using SARIMA
        sarima_forecast = forecast_sarima(time_series_data)

        # Forecast using LSTM
        lstm_forecast = forecast_lstm(time_series_data)

        # Forecast using Moving Average
        ma_forecast = forecast_moving_average(time_series_data)

        # Combine the forecasts into a single DataFrame
        combined_forecasts = pd.DataFrame({
            'Date': time_series_data['Date'],
            'SARIMA Forecast': sarima_forecast,
            'LSTM Forecast': lstm_forecast,
            'Moving Average Forecast': ma_forecast
        })

        # Save the combined forecasts to a CSV file
        combined_forecasts.to_csv('combined_forecasts.csv', index=False)
    else:
        print('No data available for the specified filters.')


In [42]:
def forecast_sarima(time_series_data):
    # SARIMA model
    model = SARIMAX(time_series_data['Value'], order=(1, 1, 1), seasonal_order=(1, 1, 1, 12), enforce_stationarity=False)
    sarima_fit = model.fit(disp=False)
    sarima_forecast = sarima_fit.get_forecast(steps=len(time_series_data)).predicted_mean
    return sarima_forecast

In [43]:
def forecast_lstm(time_series_data):
    # Normalize data for LSTM
    scaler = MinMaxScaler()
    scaled_data = scaler.fit_transform(time_series_data['Value'].values.reshape(-1, 1))

    # Prepare data for LSTM
    x_train, y_train = create_lstm_dataset(scaled_data, time_steps=3)
    
    # Build LSTM model
    lstm_model = build_lstm_model(x_train.shape[1])

    # Train LSTM model
    lstm_model.fit(x_train, y_train, epochs=50, batch_size=32, verbose=0)

    # Make predictions
    lstm_forecast = predict_lstm_forecast(lstm_model, scaled_data)

    # Inverse transform the forecasted values
    lstm_forecast = scaler.inverse_transform(lstm_forecast.reshape(-1, 1)).flatten()

    return lstm_forecast



In [44]:
def forecast_moving_average(time_series_data):
    # Moving average forecast
    ma_forecast = time_series_data['Value'].rolling(window=3).mean()
    ma_forecast = ma_forecast.fillna(0)  # Filling NaN values with 0 for simplicity
    return ma_forecast

In [45]:
def create_lstm_dataset(data, time_steps=1):
    x, y = [], []
    for i in range(len(data) - time_steps):
        a = data[i:(i + time_steps), 0]
        x.append(a)
        y.append(data[i + time_steps, 0])
    return np.array(x), np.array(y)

In [46]:
def build_lstm_model(input_shape):
    model = Sequential()
    model.add(LSTM(50, activation='relu', input_shape=(input_shape, 1)))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mse')
    return model

In [47]:
def predict_lstm_forecast(model, data):
    x_test = create_lstm_dataset(data, time_steps=3)[0]
    x_test = np.reshape(x_test, (x_test.shape[0], x_test.shape[1], 1))
    lstm_forecast = model.predict(x_test)
    return lstm_forecast

In [48]:
 #Example usage:
forecast_and_save(region='Desktop', start_year='2023', start_month='04', end_year='2023', end_month='04')

No data available for the specified filters.


#### 2(c) : From the csv created above find out which algorithm best predicted the values for the input parameters. The function should also create a summary table for the forecast where we could see the accuracy of the time series forecasting model. You may add more evaluating criteria, but the following are a must have:<br> i.	Mean Absolute Percentage Error (MAPE)<br>ii.	Mean Squared Error (MSE)


In [49]:
import pandas as pd
from sklearn.metrics import mean_absolute_error, mean_squared_error
import numpy as np

In [50]:
def evaluate_forecasts(file_path):
    # Read the combined forecasts from the CSV file
    combined_forecasts = pd.read_csv(file_path)

    # Evaluate SARIMA forecast
    sarima_mape = calculate_mape(combined_forecasts['Value'], combined_forecasts['SARIMA Forecast'])
    sarima_mse = mean_squared_error(combined_forecasts['Value'], combined_forecasts['SARIMA Forecast'])

    # Evaluate LSTM forecast
    lstm_mape = calculate_mape(combined_forecasts['Value'], combined_forecasts['LSTM Forecast'])
    lstm_mse = mean_squared_error(combined_forecasts['Value'], combined_forecasts['LSTM Forecast'])

    # Evaluate Moving Average forecast
    ma_mape = calculate_mape(combined_forecasts['Value'], combined_forecasts['Moving Average Forecast'])
    ma_mse = mean_squared_error(combined_forecasts['Value'], combined_forecasts['Moving Average Forecast'])

    # Create a summary table
    summary_table = pd.DataFrame({
        'Model': ['SARIMA', 'LSTM', 'Moving Average'],
        'MAPE': [sarima_mape, lstm_mape, ma_mape],
        'MSE': [sarima_mse, lstm_mse, ma_mse]
    })

    print("Summary Table:")
    print(summary_table)

def calculate_mape(actual, forecast):
    return np.mean(np.abs((actual - forecast) / actual)) * 100


# AB_Test Sheet questions

#### Q1) a.Have we reached the required amount of sample size to conclude the test?

In [52]:
import scipy.stats as stats

# Given parameters
MDE = 0.03  # 3%
alpha = 0.05  # Significance level
power = 0.8  # Statistical power

# Baseline conversion rate (p1)
baseline_conversion = 0.1  # Replace with the actual baseline conversion rate

# Expected conversion rate with treatment (p2)
expected_conversion = baseline_conversion + MDE

# Z-score for the specified significance level (α/2)
z_alpha_2 = stats.norm.ppf(1 - alpha/2)

# Z-score for the specified statistical power (1−β)
z_beta = stats.norm.ppf(power)

# Calculate sample size
sample_size = ((z_alpha_2 + z_beta)**2 * (baseline_conversion * (1 - baseline_conversion) + expected_conversion * (1 - expected_conversion))) / ((baseline_conversion - expected_conversion)**2)

print(f"Required Sample Size: {sample_size:.2f}")


Required Sample Size: 1771.23


### No, we have not reached the sample size.


#### (b): 2.	At 95% confidence can you tell us <br> a.	Has the test reached statistical significance?


In [54]:
import pandas as pd
import scipy.stats as stats

# Assuming your DataFrame is named 'ab_test_data'
# Replace this with the actual name of your DataFrame

# Extract the control and treatment groups from your DataFrame
control_group = ab_test_data['Control'].tolist()
treatment_group = ab_test_data['Treatment'].tolist()

# Calculate the p-value using a two-sample t-test
t_stat, p_value = stats.ttest_ind(control_group, treatment_group)

# Set the significance level
alpha = 0.05

# Check for statistical significance
if p_value < alpha:
    print(f"The test is statistically significant at {100 * (1 - alpha):.2f}% confidence.")
else:
    print(f"The test is not statistically significant at {100 * (1 - alpha):.2f}% confidence.")


NameError: name 'ab_test_data' is not defined